06_memory/module/module_malloc.c

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/vmalloc.h>
+
+enum allocator {
+	KMALLOC = 0,
+	KZALLOC,
+	VMALLOC,
+	GET_FREE_PAGES,
+};
+
+void *call_alloc(enum allocator alloc, size_t i)
+{
+	void *p = NULL;
+
+	switch (alloc) {
+	case KMALLOC:
+		p = kmalloc(i, GFP_KERNEL);
+		break;
+	case KZALLOC:
+		p = kzalloc(i, GFP_KERNEL);
+		break;
+	case VMALLOC:
+		p = vmalloc(i);
+		break;
+	case GET_FREE_PAGES:
+		p = (void *)__get_free_pages(GFP_KERNEL, i);
+		break;
+	default:
+		return NULL;
+	}
+
+	return p;
+}
+
+void call_free(enum allocator alloc, void *p, unsigned long i)
+{
+	switch (alloc) {
+	case KMALLOC:
+	case KZALLOC:
+		kfree(p);
+		break;
+	case VMALLOC:
+		vfree(p);
+		break;
+	case GET_FREE_PAGES:
+		free_pages((unsigned long)p, i);
+		break;
+	default:
+		break;
+	}
+}
+
+void measure_one(enum allocator alloc, size_t i)
+{
+	ktime_t t[3];
+	struct timespec64 t_a, t_f;
+	void *p;
+	size_t buffer_allocated = i;
+
+	t[0] = ktime_get();
+
+	p = call_alloc(alloc, i);
+	if (p == NULL)
+		return;
+
+	t[1] = ktime_get();
+	call_free(alloc, p, i);
+	t[2] = ktime_get();
+
+	t_a = ktime_to_timespec64(ktime_sub(t[1], t[0]));
+	t_f = ktime_to_timespec64(ktime_sub(t[2], t[1]));
+
+	if (alloc == GET_FREE_PAGES)
+		buffer_allocated = i * PAGE_SIZE;
```

Comment:
Is this conversation correct?

Reply:
Agree, fixed in https://github.com/Kernel-GL-HRK/gl_kernel_procamp_2021/pull/148/commits/7181b5d0c66692b76dc01a72d39a7497a761e93c